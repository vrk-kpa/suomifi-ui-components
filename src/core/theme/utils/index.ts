export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
