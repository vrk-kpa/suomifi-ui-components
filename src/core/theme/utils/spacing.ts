import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (props: SpacingProps | undefined): string => {
  if (!props) return '';

  const cssStyles = Object.entries(props)
    .map(([key, value]) =>
      getCSSSpacing(defaultSuomifiTheme, key as keyof SpacingProps, value),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return `margin-right: ${amount}; margin-left: ${amount};`;
    case 'my':
      return `margin-top: ${amount}; margin-bottom: ${amount};`;
    case 'px':
      return `padding-right: ${amount}; padding-left: ${amount};`;
    case 'py':
      return `padding-top: ${amount}; padding-bottom: ${amount};`;
    default:
      return `${[cssSelector[key]]}: ${amount};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
