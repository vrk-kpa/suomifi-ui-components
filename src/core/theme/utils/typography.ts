import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
