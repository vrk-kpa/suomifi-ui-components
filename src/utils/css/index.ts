export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
