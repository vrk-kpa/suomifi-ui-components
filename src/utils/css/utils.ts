import { css, FlattenSimpleInterpolation } from 'styled-components';
import { ValueUnit } from 'suomifi-design-tokens';
export { FlattenSimpleInterpolation };

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

/**
 * Return FlattenSimpleInterpolation compatible string or number
 * @param cssValue compatible value or {value, unit} pair-object
 */
const cssValueToString = (cssValue: number | string | ValueAndUnit) => {
  if (!!cssValue && typeof cssValue === 'object' && 'value' in cssValue) {
    const { value, unit } = cssValue;
    return !!unit ? `${value}${unit}` : value;
  }
  return cssValue;
};

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, m => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 */
export const cssObjectToCss = <T>(value: T) => css`
  ${Object.entries(value)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
