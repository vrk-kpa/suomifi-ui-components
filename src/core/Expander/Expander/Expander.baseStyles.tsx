import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { element, font } from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${element(theme)}
  ${font(theme)('bodyText')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  background-color: ${theme.colors.whiteBase};
  position: relative;
  padding: 0;
  border: 1px solid ${theme.colors.highlightBase};
  border-radius: ${theme.radiuses.basic};
  width: 100%;
  max-width: 100%;

  &.fi-expander {
    display: block;
  }
  &:before {
    background-color: ${theme.colors.highlightLight4};
    opacity: 0;
  }

  &.fi-expander--open {
    border-bottom: none;
  }
`;
