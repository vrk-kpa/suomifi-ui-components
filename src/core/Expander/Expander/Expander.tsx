import React, { Component, ReactNode, forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { AutoId } from '../../utils/AutoId/AutoId';
import { HtmlDivWithRef, HtmlDivWithRefProps } from '../../../reset';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import {
  ExpanderGroupConsumer,
  ExpanderGroupProviderState,
} from '../ExpanderGroup/ExpanderGroup';
import { baseStyles } from './Expander.baseStyles';
import { filterDuplicateKeys } from '../../../utils/common/common';

const baseClassName = 'fi-expander';
const openClassName = `${baseClassName}--open`;

export interface ExpanderProviderState {
  /** Callback for communicating ExpanderTitle button event to Expander  */
  onToggleExpander: () => void;
  /** Open state for expander */
  open: boolean;
  /** Id for expander button */
  titleId: string | undefined;
  /** Id for expander content */
  contentId: string | undefined;
}

const defaultProviderValue: ExpanderProviderState = {
  onToggleExpander: () => null,
  open: false,
  titleId: undefined,
  contentId: undefined,
};

const { Provider: ExpanderProvider, Consumer: ExpanderConsumer } =
  React.createContext(defaultProviderValue);

export interface ExpanderProps extends MarginProps, HtmlDivWithRefProps {
  /**
   * Children extend type ExpanderTitleBaseProps or ExpanderContentBaseProps
   * ExpanderProviderState context is used to communicate between title, content and expander
   */
  children: ReactNode;
  /** CSS class for custom styles */
  className?: string;
  /**
   * HTML id for Expander, must be unique. Duplicate id's break ExpanderGroup functionality.
   * If no id is provided, one will be generated automatically
   */
  id?: string;
  /** Initial open state
   * @default false
   */
  defaultOpen?: boolean;
  /** Controlled open property */
  open?: boolean;
  /** Callback fired on click */
  onOpenChange?: (open: boolean) => void;
  /** Ref is placed to the outermost div element of the component. Alternative for React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLDivElement>;
}

export interface ExpanderTitleBaseProps {
  /** CSS class for custom styles */
  className?: string;
  /** Expander consumer for open state and toggle open callback */
  consumer: ExpanderProviderState;
}

export interface ExpanderContentBaseProps {
  /** CSS class for custom styles */
  className?: string;
  /** Expander consumer for open state */
  consumer: ExpanderProviderState;
}

interface BaseExpanderProps extends ExpanderProps {
  id: string;
  consumer: ExpanderGroupProviderState;
}

class BaseExpander extends Component<BaseExpanderProps & SuomifiThemeProp> {
  state: ExpanderState = {
    openState: this.props.defaultOpen || false,
  };

  constructor(props: BaseExpanderProps & SuomifiThemeProp) {
    super(props);
    if (!!props.id) {
      const defaultOpen =
        props.open !== undefined ? props.open : props.defaultOpen || false;
      props.consumer.onExpanderOpenChange(props.id, defaultOpen);
    }
  }

  componentDidUpdate(prevProps: BaseExpanderProps, prevState: ExpanderState) {
    const { consumer, open } = this.props;
    const controlled = open !== undefined;
    const currentState = controlled ? !!open : this.state.openState;
    // update group state when id changes
    if (prevProps.id !== this.props.id) {
      if (!!prevProps.id) {
        consumer.onExpanderOpenChange(prevProps.id, undefined);
      }
      consumer.onExpanderOpenChange(this.props.id, currentState);
    }
    // handle consumer open change event
    if (
      consumer.expanderGroupOpenState !==
      prevProps.consumer?.expanderGroupOpenState
    ) {
      if (
        (!controlled &&
          !!this.state.openState !==
            consumer.expanderGroupOpenState.targetOpenState) ||
        (controlled && open !== consumer.expanderGroupOpenState.targetOpenState)
      ) {
        this.handleOpenChange();
      }
    }
    // handle expander open change event
    if (
      (!controlled && prevState.openState !== this.state.openState) ||
      (controlled && prevProps.open !== open)
    ) {
      if (!!this.props.id) {
        consumer.onExpanderOpenChange(this.props.id, currentState);
      }
    }
  }

  componentWillUnmount() {
    this.props.consumer.onExpanderOpenChange(this.props.id, undefined);
  }

  handleOpenChange = () => {
    const { open, onOpenChange } = this.props;
    const { openState } = this.state;
    const controlled = open !== undefined;
    const newOpenState = controlled ? !!open : !openState;
    if (!controlled) {
      this.setState({ openState: newOpenState });
    }
    if (!!onOpenChange) {
      onOpenChange(newOpenState);
    }
  };

  render() {
    const {
      id,
      open,
      defaultOpen,
      onOpenChange,
      className,
      theme,
      children,
      consumer,
      style,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);
    const openState = open !== undefined ? !!open : this.state.openState;

    return (
      <HtmlDivWithRef
        id={id}
        {...passProps}
        className={classnames(className, baseClassName, {
          [openClassName]: !!openState,
        })}
        style={style}
      >
        <ExpanderProvider
          value={{
            open: openState,
            contentId: `${id}_content`,
            titleId: `${id}_title`,
            onToggleExpander: this.handleOpenChange,
          }}
        >
          {children}
        </ExpanderProvider>
      </HtmlDivWithRef>
    );
  }
}

const StyledExpander = styled(
  (props: BaseExpanderProps & SuomifiThemeProp & GlobalMarginProps) => {
    const { globalMargins, ...passProps } = props;
    return <BaseExpander {...passProps} />;
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.dropdown,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

interface ExpanderState {
  openState: boolean;
}

const Expander = forwardRef(
  (props: ExpanderProps, ref: React.Ref<HTMLDivElement>) => {
    const { id: propId, ...passProps } = props;
    return (
      <SpacingConsumer>
        {({ margins }) => (
          <SuomifiThemeConsumer>
            {({ suomifiTheme }) => (
              <AutoId id={propId}>
                {(id) => (
                  <ExpanderGroupConsumer>
                    {(consumer) => (
                      <StyledExpander
                        theme={suomifiTheme}
                        globalMargins={margins}
                        id={id}
                        forwardedRef={ref}
                        {...passProps}
                        consumer={consumer}
                      />
                    )}
                  </ExpanderGroupConsumer>
                )}
              </AutoId>
            )}
          </SuomifiThemeConsumer>
        )}
      </SpacingConsumer>
    );
  },
);

Expander.displayName = 'Expander';
export { Expander, ExpanderConsumer, ExpanderProvider };
