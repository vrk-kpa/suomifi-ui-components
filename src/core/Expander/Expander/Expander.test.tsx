import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { axeTest } from '../../../utils/test';
import { Expander, ExpanderProps } from './Expander';
import {
  ExpanderContent,
  ExpanderTitleButton,
  ExpanderTitleButtonProps,
} from '../';

describe('Basic expander', () => {
  const TestExpanderWithProps = (titleProps: ExpanderTitleButtonProps) => (
    <Expander className="expander-test">
      <ExpanderTitleButton {...titleProps} data-testid="expander-title" />
      <ExpanderContent>Test expander content</ExpanderContent>
    </Expander>
  );

  const TestExpander = TestExpanderWithProps({
    children: 'Test expander',
  });

  it('render with the same component on the same container does not remount', () => {
    const { getByTestId, rerender } = render(TestExpander);
    expect(getByTestId('expander-title').textContent).toBe('Test expander');

    // re-render the same component with different props
    rerender(
      TestExpanderWithProps({
        children: 'Test expander two',
      }),
    );
    expect(getByTestId('expander-title').textContent).toBe('Test expander two');
  });

  it('shoud match snapshot', () => {
    const expanderRenderer = render(TestExpander);
    const { container } = expanderRenderer;
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', axeTest(TestExpander));
});

describe('Custom id', () => {
  const ExpanderWithCustomId = (
    <Expander id="test-id" {...{ 'data-testid': 'expander-custom-id' }}>
      <ExpanderTitleButton
        toggleButtonProps={{ 'data-testid': 'expander-custom-id-title' }}
      >
        Test expander with custom id
      </ExpanderTitleButton>
      <ExpanderContent {...{ 'data-testid': 'expander-custom-id-content' }}>
        Test expander with custom id content
      </ExpanderContent>
    </Expander>
  );

  it('is passed on to expander', () => {
    const { getByTestId } = render(ExpanderWithCustomId);
    const div = getByTestId('expander-custom-id');
    expect(div).toHaveAttribute('id', 'test-id');
  });
});

describe('defaultOpen', () => {
  const DefaultOpenExpander = (props?: Omit<ExpanderProps, 'children'>) => (
    <Expander {...props} defaultOpen={true}>
      <ExpanderTitleButton
        {...{ 'data-testid': 'expander-open-by-default-title' }}
      >
        Test expander open by default
      </ExpanderTitleButton>
      <ExpanderContent>Test expander open by default content</ExpanderContent>
    </Expander>
  );

  it('classnames will be removed when clicked', () => {
    const mockClickHandler = jest.fn();
    const { getByTestId, getByRole } = render(
      DefaultOpenExpander({ onOpenChange: mockClickHandler }),
    );
    const buttonToClick = getByRole('button');
    const titleDiv = getByTestId('expander-open-by-default-title');
    fireEvent.click(buttonToClick);
    expect(titleDiv).toHaveClass('fi-expander_title-button');
    expect(titleDiv).not.toHaveClass('fi-expander_title-button--open');
  });
});

describe('onOpenChange', () => {
  it('is called', async () => {
    const mockClickHandler = jest.fn();
    const { getByRole } = render(
      <Expander onOpenChange={mockClickHandler}>
        <ExpanderTitleButton>Test expander open by default</ExpanderTitleButton>
        <ExpanderContent>Test expander open by default content</ExpanderContent>
      </Expander>,
    );
    const button = getByRole('button');
    fireEvent.click(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
  });
});

describe('open', () => {
  const ControlledExpander = (props?: Partial<ExpanderProps>) => (
    <Expander className="expander-onclick-test" open={true} {...props}>
      <ExpanderTitleButton {...{ 'data-testid': 'expander-title-id' }}>
        Test expander onClick testing
      </ExpanderTitleButton>
      <ExpanderContent>Test expander click testing content</ExpanderContent>
    </Expander>
  );

  it('open-classnames should be found ', async () => {
    const { getByTestId } = render(ControlledExpander());
    const div = getByTestId('expander-title-id');
    expect(div).toHaveClass('fi-expander_title-button--open');
  });

  it('is clicked. Should not change as it is controlled outside', async () => {
    const mockClickHandler = jest.fn();
    const { getByRole, getByTestId } = render(
      ControlledExpander({ onOpenChange: mockClickHandler }),
    );
    const button = getByRole('button');
    fireEvent.click(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
    const div = getByTestId('expander-title-id');
    expect(div).toHaveClass('fi-expander_title-button--open');
  });
});

describe('margin prop', () => {
  it('should have margin style from margin prop', () => {
    const { getByTestId } = render(
      <Expander data-testid="expander-test" margin="xs">
        <ExpanderContent>Test expander content</ExpanderContent>
      </Expander>,
    );
    const div = getByTestId('expander-test');
    expect(div).toHaveStyle('margin: 10px');
  });

  it('should have margin style overridden by style prop', async () => {
    const { container } = render(
      <Expander data-testid="expander-test" margin="l" style={{ margin: 2 }}>
        <ExpanderContent>Test expander content</ExpanderContent>
      </Expander>,
    );
    expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
  });
});
