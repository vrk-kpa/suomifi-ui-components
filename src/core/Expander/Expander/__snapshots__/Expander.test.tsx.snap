// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic expander shoud match snapshot 1`] = `
.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c5:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c5::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c5::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c5::-webkit-inner-spin-button {
  height: auto;
}

.c5::-webkit-outer-spin-button {
  height: auto;
}

.c5 ::before,
.c5 ::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2 ::before,
.c2 ::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4 ::before,
.c4 ::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  background-color: hsl(0, 0%, 100%);
  position: relative;
  padding: 0;
  border: 1px solid hsl(212, 63%, 45%);
  border-radius: 2px;
  width: 100%;
  max-width: 100%;
}

.c1.fi-expander {
  display: block;
}

.c1:before {
  background-color: hsl(212, 63%, 98%);
  opacity: 0;
}

.c1.fi-expander--open {
  border-bottom: none;
}

.c7 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  background-color: hsl(0, 0%, 100%);
  border-radius: inherit;
  position: relative;
  visibility: hidden;
  height: 0;
  overflow: hidden;
  word-break: break-word;
  transform-origin: top;
}

.c7.fi-expander {
  display: block;
}

.c7:not(.fi-expander_content--no-padding) {
  padding: 0 16px;
}

.c7.fi-expander_content--open {
  visibility: visible;
  height: auto;
  border-bottom: 1px solid hsl(212, 63%, 45%);
  border-top: 1px solid transparent;
  border-top-left-radius: 0;
  border-top-right-radius: 0;
}

.c7.fi-expander_content--open:not(.fi-expander_content--no-padding) {
  padding: 10px 20px 20px;
}

.c6 {
  vertical-align: baseline;
}

.c6.fi-icon {
  display: inline-block;
}

.c6 .fi-icon-base-fill {
  fill: currentColor;
}

.c6 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c6.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c6.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c3 {
  color: hsl(0, 0%, 13%);
  position: relative;
  width: 100%;
  max-width: 100%;
  min-height: 60px;
  background-color: hsl(212, 63%, 98%);
  border-radius: inherit;
}

.c3:hover {
  background: linear-gradient(0deg, hsl(215, 100%, 95%), hsl(215, 100%, 97%));
}

.c3 .fi-expander_title-button {
  display: block;
}

.c3.fi-expander_title-button--open {
  background-color: hsl(0, 0%, 100%);
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

.c3 .fi-expander_title-button_button {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: calc(20 / 14);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(212, 63%, 45%);
  display: inline-block;
  width: 100%;
  max-width: 100%;
  min-height: 60px;
  padding: 17px 60px 16px 20px;
}

.c3 .fi-expander_title-button_button:focus {
  outline: 0;
}

.c3 .fi-expander_title-button_button:focus-within {
  outline: 3px solid transparent;
}

.c3 .fi-expander_title-button_button:focus-within:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0, 0%, 100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196, 77%, 44%);
  z-index: 9999;
}

.c3 .fi-expander_title-button_button :not(:focus-visible) {
  outline: 0;
}

.c3 .fi-expander_title-button_button :not(:focus-visible):after {
  content: none;
}

.c3 .fi-expander_title-button_button * {
  cursor: pointer;
}

.c3 .fi-expander_title-button_button:hover,
.c3 .fi-expander_title-button_button:active,
.c3 .fi-expander_title-button_button:focus,
.c3 .fi-expander_title-button_button:focus-within {
  cursor: pointer;
}

.c3 .fi-expander_title-button-icon {
  position: absolute;
  top: 0;
  right: 0;
  height: 20px;
  width: 20px;
  margin: 20px;
}

.c3 .fi-expander_title-button--open .fi-expander_title-button-icon,
.c3 .fi-expander_title-button-icon--open {
  transform: rotate(-180deg);
}

<div
  class="c0 c1 expander-test fi-expander"
  id="4"
>
  <div
    class="c2 c3 fi-expander_title-button"
    data-testid="expander-title"
  >
    <span
      class="c4"
    >
      <button
        aria-controls="4_content"
        aria-expanded="false"
        class="c5 fi-expander_title-button_button"
        type="button"
      >
        <span
          class="c4"
          id="4_title"
        >
          Test expander
        </span>
        <svg
          aria-hidden="true"
          class="fi-icon c6 fi-expander_title-button-icon"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="fi-icon-base-fill"
            d="M12.99 15.61c-.273.26-.632.39-.99.39s-.717-.13-.99-.39l-5.6-5.334a1.287 1.287 0 0 1 0-1.885 1.448 1.448 0 0 1 1.98 0l4.61 4.39 4.61-4.39a1.448 1.448 0 0 1 1.98 0c.547.52.547 1.365 0 1.885l-5.6 5.333Z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </span>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="4_title"
    class="c2 c7 fi-expander_content"
    id="4_content"
    role="region"
  >
    Test expander content
  </div>
</div>
`;
